### PR TITLE
Enhancements in EC2 and Added Support for AMIs

### PR DESCRIPTION
**Pull request by kedarghule**
This PR does the following:

1) Currently the auto_scaling_groups.py code does not use the UNWIND approach, instead it uses loops to populate the AWS resources. This PR rectifies that. Additionally, some more attributes of AutoScalingGroups are populated.

2) Some additional useful attributes missing from EC2 Instances and Subnets get populated too! Corresponding changes have been made in the documentation. (I specifically didn't add the UNWIND approach here because I wasn't quite sure of the code structure here and I didn't want to break things 🙈)

3) The documentation for Auto Scaling Groups was missing all this while. Added it!

4) And the one major update in this PR is - Added support for Amazon Machine Images (AMIs) - images.py in the ec2 folder.

### CODE REVIEW

File: docs/schema/aws.md [hunk 1]
```diff
@@ -2158,3 +2174,80 @@ Representation of an AWS [API Gateway Resource](https://docs.aws.amazon.com/apig
...
+        ```
+- AWS EC2 Instances are members of one or more AWS Auto Scaling Groups.
+
+        ```
+        (EC2Instance)-[MEMBER_AUTO_SCALE_GROUP]->(AutoScalingGroup)
+        ```
+
+## EC2Image
+
+Representation of an AWS [EC2 Images (AMIs)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html).
```
Comment (achantavy) [inline docs/schema/aws.md:2222]: Let's also link to the API doc: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeImages.html

File: docs/schema/aws.md [hunk 2]
```diff
@@ -2158,3 +2174,80 @@ Representation of an AWS [API Gateway Resource](https://docs.aws.amazon.com/apig
         ```
         (APIGatewayRestAPI)-[RESOURCE]->(APIGatewayResource)
         ```
+
+## AutoScalingGroup
+
+Representation of an AWS [Auto Scaling Group Resource](https://docs.aws.amazon.com/autoscaling/ec2/userguide/AutoScalingGroup.html).
```
Comment (achantavy) [inline docs/schema/aws.md:2180]: Thanks for including this. Can you also add a link to the AWS API doc?

File: cartography/intel/aws/ec2/images.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return images
+
+
+@timeit
+def load_images(
+        neo4j_session: neo4j.Session, data: List[Dict], region: str, current_aws_account_id: str, update_tag: int,
+) -> None:
+    ingest_images = """
+    UNWIND {images_list} as image
+        MERGE (i:EC2Image{id: image.ImageId})
```
Comment (achantavy) [inline cartography/intel/aws/ec2/images.py:35]: Is the ImageId guaranteed to be unique across all AWS like an ARN?

**Review by achantavy — COMMENTED** [review]
Left a couple of questions but it's looking really good! Thanks for taking this on.

File: cartography/intel/aws/ec2/auto_scaling_groups.py [hunk 4]
```diff
@@ -26,88 +26,94 @@ def get_ec2_auto_scaling_groups(boto3_session: boto3.session.Session, region: st
...
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {update_tag}
+        WITH instance
+        MATCH (aa:AWSAccount{id: {AWS_ACCOUNT_ID}})
+        MERGE (aa)-[r:RESOURCE]->(instance)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {update_tag}
     """
 
+    # neo4j does not accept datetime objects and values. This loop is used to convert
```
Comment (achantavy) [inline cartography/intel/aws/ec2/auto_scaling_groups.py:79]: Nit: technically not correct anymore as of 3.5 iirc, but it's fine to keep as `str()`.

File: cartography/intel/aws/ec2/images.py (hunk 3 above)
Comment (achantavy) [reply cartography/intel/aws/ec2/images.py:35]: If ImageId is only unique per region, let's set `i.id` to be `{ImageId}|{region}` and set `i.imageid = {ImageId}`. Also add an index for `i.imageid`.

File: cartography/intel/aws/ec2/auto_scaling_groups.py (hunk 4 above)
Comment (kedarghule) [reply cartography/intel/aws/ec2/auto_scaling_groups.py:79]: Ah I see. Okay, I'll just remove that comment.

File: docs/schema/aws.md (hunk 1 above)
Comment (kedarghule) [reply docs/schema/aws.md:2222]: Done

File: docs/schema/aws.md (hunk 2 above)
Comment (kedarghule) [reply docs/schema/aws.md:2180]: Done

File: cartography/intel/aws/ec2/images.py (hunk 3 above)
Comment (kedarghule) [reply cartography/intel/aws/ec2/images.py:35]: cc: @mpurusottamc

Comment (mpurusottamc) [reply cartography/intel/aws/ec2/images.py:35]: @achantavy AMI ID are unique per region. So, we will follow your suggestion of appending region to the id.
[Finding an AMI](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/finding-an-ami.html#finding-an-ami-console)
Thanks @kedarghule

Comment (kedarghule) [reply cartography/intel/aws/ec2/images.py]: @achantavy and @mpurusottamc  Changes made and committed 🚀

**Comment by kedarghule** [comment]
@achantavy All suggested changes as per your review are done here. Anymore changes required? :)

**Review by achantavy — APPROVED** [review]
Very much appreciate the contribution as always @kedarghule  :D

**Pull request merged**